### ISSUE TITLE
Stop casting SQLDataType.OTHER to VARCHAR

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11511]
A very historic piece of code in `DefaultBinding` casts the `OTHER` data type to `VARCHAR` because at least Db2 and Derby cannot have `NULL` values without an explicit data type.

Times have changed and jOOQ is used as a parser more and more frequently. In a parser context, we may not know the data type of every bind variable, and we also do not have a bind value associated with it at parse time (unlike when people use jOOQ as a DSL), so the logic that prevents jOOQ from generating bad SQL in Db2 and Derby gets in the way in various other dialects.

For example, this line of code:

```java
System.out.println(DSL
    .using(SQLDialect.POSTGRES)
    .parser()
    .parseQuery("select 1 from t where a = ?")
    .getSQL(ParamType.INDEXED)
);
```

Produces the following output:

```sql
select 1 from t where a = cast(? as varchar)
```

Which is quite wrong whenever `t.a` is not a textual data type:

- We don't know the type of `t.a`, so it is `OTHER` (unless users use meta lookups, but this isn't working correctly for all cases yet: https://github.com/jOOQ/jOOQ/issues/11061)
- We don't have a bind value as an example of what could be bound to `?`, so we cannot infer any data types from that

In the latter case:

```java
System.out.println(DSL
    .using(SQLDialect.POSTGRES)
    .parser()
    .parseQuery("select 1 from t where a = ?", 1) // Example bind value
    .getSQL(ParamType.INDEXED)
);
```

We're no longer generating the bad SQL, but this, instead:

```sql
select 1 from t where a = ?
```

So, this can act as a workaround

**Comment by lukaseder** [comment]
While the results will be better, and this fix will also fix #5851 as a side effect, and has exposed a related problem with possibly wrong casts in `JSON_OBJECT()` and `JSON_ARRAY()`, I'm reluctant to backport it to 3.14.8 or any 3.14 patch release, as such a change can have significant, yet subtle impact on client code that may have worked with the assumption of this cast being made by jOOQ.

It is, however, a very important fix for users that use jOOQ as a parser/translator, especially when using the `ParsingConnection`

**Comment by lukaseder** [comment]
This change will create regressions that are difficult to fix in these two cases:

- https://github.com/jOOQ/jOOQ/issues/1028
- https://github.com/jOOQ/jOOQ/issues/1029

The question is whether these are worth fixing, i.e. do end users really project named parameters:

- without associated data type
- without associated value
- and find the existing `VARCHAR` cast acceptable rather than clever and possibly wrong?

Perhaps it's better to create a situation where an error occurs instead of just casting the value to a random type like `VARCHAR`. But these regressions are proof that we can't backport this fix.

**Comment by lukaseder** [comment]
The issues mentioned above  run tests like this:

```java
public void testUnknownBindTypes() throws Exception {

    // [#1028] [#1029] Named params without any associated type information
    Select<?> select = create()
        .select(
            param("p1"),
            param("p2"),
            param())
        .where(param().eq(1));

    select.bind(1, "10");
    select.bind(2, null);
    select.bind(3, "A");
    select.bind(4, 1);
    Result<?> result3 = select.fetch();

    assertEquals(1, result3.size());
    assertEquals("10", result3.getValue(0, 0));
    assertEquals(null, result3.getValue(0, 1));
    assertEquals("A", result3.getValue(0, 2));
}
```

Placing a `Param` marker without type *nor* value makes sense only when we use `Query.bind()` later on. Instead of mutating the `param`, `Query.bind()` could replace the `Param` in the context of a wrapping `Query` execution (see https://github.com/jOOQ/jOOQ/issues/11513), which would embrace immutability *and* improve performance of the `bind` operation, *and* fix this new regression introduced by us stopping the cast. Which was incorrect in this case anyway, if we hadn't used string values in the first place.

Hence, I'm ignoring the regression on #1028 and #1029 for now, as there exist workarounds (making the type explicit with `param(name, type)`, and the current test will be fixed via #11513, whose priority I'll increase.
